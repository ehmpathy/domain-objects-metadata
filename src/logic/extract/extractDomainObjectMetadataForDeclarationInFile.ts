import { ClassDeclaration, isInterfaceDeclaration, SourceFile, SyntaxKind } from 'typescript';

import { DomainObjectMetadata, DomainObjectVariant } from '../../domain/objects/DomainObjectMetadata';
import { extractClassExtendsWhat } from './extractClassExtendsWhat';
import { extractPropertiesFromInterfaceDeclaration } from './extractPropertiesFromInterfaceDeclaration';
import { extractRelevantStaticPropertiesFromClassDeclaration } from './extractRelevantStaticPropertiesFromClassDeclaration';
import { isAClassDecorationWhichExtendsDomainObject } from './isAClassDeclarationWhichExtendsDomainObject';

export const extractDomainObjectMetadataForDeclarationInFile = ({
  classDeclaration,
  file,
}: {
  classDeclaration: ClassDeclaration;
  file: SourceFile;
}): DomainObjectMetadata => {
  // sanity check that it _is_ a declaration for domain object
  if (!isAClassDecorationWhichExtendsDomainObject(classDeclaration))
    throw new Error('called extraction of domain object metadata on a class declaration which does not extend DomainObject');

  // grab the name of the class
  if (!classDeclaration.name?.text)
    throw new Error(`domain objects must be given names in their definition. found one without name in ${file.fileName}`);
  const name = classDeclaration.name.text;

  // grab which class it extends
  const extendsWhat: DomainObjectVariant = extractClassExtendsWhat(classDeclaration) as DomainObjectVariant;

  // grab the name of the interface it implements
  const implementsClause = classDeclaration.heritageClauses!.find((clause) => clause.token === SyntaxKind.ImplementsKeyword);
  if (!implementsClause) throw new Error(`domain objects must implement an interface which defines their properties. not found for ${name}`);
  const nameOfImplementedInterface = (implementsClause.types[0].expression as any).escapedText;

  // find the interface that it implements in the file
  const interfaceDeclaration = file.statements
    .filter(isInterfaceDeclaration)
    .find((declaration) => declaration.name.text === nameOfImplementedInterface);
  if (!interfaceDeclaration)
    throw new Error(
      `domain object class-declarations must be in the same file as their 'implements' interface-declarations. not satisfied for ${name}`,
    );

  // extract the properties of the class, defined on the interface
  const properties = extractPropertiesFromInterfaceDeclaration(interfaceDeclaration);

  // grab the static properties of the class: `unique` and `updatable` specifically
  const relevantStaticProperties = extractRelevantStaticPropertiesFromClassDeclaration(classDeclaration);

  // define the decorations of this domain-object
  const decorations = {
    unique: relevantStaticProperties.unique,
    updatable: relevantStaticProperties.updatable,
  };

  // return the domain object metadata
  return new DomainObjectMetadata({
    name,
    extends: extendsWhat,
    properties,
    decorations,
  });
};
