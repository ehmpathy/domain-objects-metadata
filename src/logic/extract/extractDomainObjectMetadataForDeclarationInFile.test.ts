import ts, { isClassDeclaration } from 'typescript';

import { extractDomainObjectMetadataForDeclarationInFile } from './extractDomainObjectMetadataForDeclarationInFile';

describe('extractDomainObjectMetadataForDeclarationInFile', () => {
  it('should be able to extract domain object metadata for simple value object', () => {
    const program = ts.createProgram(
      [`${__dirname}/../__test_assets__/Address.ts`],
      {},
    );
    const file = program
      .getSourceFiles()
      .find((thisFile) => thisFile.fileName.includes('/Address.ts'))!; // grab the right file
    const classDeclaration = file.statements.find(isClassDeclaration)!;
    const metadata = extractDomainObjectMetadataForDeclarationInFile({
      file,
      classDeclaration,
    });
    // console.log(JSON.stringify(metadata, null, 2));

    // log an example
    expect(metadata).toMatchSnapshot();
  });
  it('should be able to extract domain object metadata for a simple entity', () => {
    const program = ts.createProgram(
      [`${__dirname}/../__test_assets__/LoadingDock.ts`],
      {},
    );
    const file = program
      .getSourceFiles()
      .find((thisFile) => thisFile.fileName.includes('/LoadingDock.ts'))!; // grab the right file
    const classDeclaration = file.statements.find(isClassDeclaration)!;
    const metadata = extractDomainObjectMetadataForDeclarationInFile({
      file,
      classDeclaration,
    });
    // console.log(JSON.stringify(metadata, null, 2));

    // log an example
    expect(metadata).toMatchSnapshot();
  });
  it('should be able to extract domain object metadata for a more complicated entity', () => {
    const program = ts.createProgram(
      [`${__dirname}/../__test_assets__/Item.ts`],
      {},
    );
    const file = program
      .getSourceFiles()
      .find((thisFile) => thisFile.fileName.includes('/Item.ts'))!; // grab the right file
    const classDeclaration = file.statements.find(isClassDeclaration)!;
    const metadata = extractDomainObjectMetadataForDeclarationInFile({
      file,
      classDeclaration,
    });
    // console.log(JSON.stringify(metadata, null, 2));

    // log an example
    expect(metadata).toMatchSnapshot();
  });
});
