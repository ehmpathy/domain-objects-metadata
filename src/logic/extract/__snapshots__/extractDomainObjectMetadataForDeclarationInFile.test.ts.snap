// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractDomainObjectMetadataForDeclarationInFile should be able to extract domain object metadata for a more complicated entity 1`] = `
DomainObjectMetadata {
  "decorations": Object {
    "unique": Array [
      "upc",
    ],
    "updatable": Array [
      "colors",
      "quantity",
      "description",
    ],
  },
  "extends": "DomainEntity",
  "name": "Item",
  "properties": Object {
    "colors": DomainObjectProperty {
      "nullable": false,
      "of": DomainObjectProperty {
        "nullable": false,
        "required": true,
        "type": "STRING",
      },
      "required": true,
      "type": "ARRAY",
    },
    "description": DomainObjectProperty {
      "nullable": false,
      "required": true,
      "type": "STRING",
    },
    "price": DomainObjectProperty {
      "nullable": false,
      "required": true,
      "type": "NUMBER",
    },
    "upc": DomainObjectProperty {
      "nullable": false,
      "required": true,
      "type": "STRING",
    },
  },
}
`;

exports[`extractDomainObjectMetadataForDeclarationInFile should be able to extract domain object metadata for a simple entity 1`] = `
DomainObjectMetadata {
  "decorations": Object {
    "unique": Array [
      "gate",
    ],
    "updatable": Array [
      "operational",
      "phoneNumber",
    ],
  },
  "extends": "DomainEntity",
  "name": "LoadingDock",
  "properties": Object {
    "gateNumber": DomainObjectProperty {
      "nullable": false,
      "required": true,
      "type": "NUMBER",
    },
    "id": DomainObjectProperty {
      "nullable": false,
      "required": false,
      "type": "NUMBER",
    },
    "operational": DomainObjectProperty {
      "nullable": false,
      "required": true,
      "type": "BOOLEAN",
    },
    "phoneNumber": DomainObjectProperty {
      "nullable": true,
      "required": true,
      "type": "STRING",
    },
    "uuid": DomainObjectProperty {
      "nullable": false,
      "required": false,
      "type": "STRING",
    },
  },
}
`;

exports[`extractDomainObjectMetadataForDeclarationInFile should be able to extract domain object metadata for simple value object 1`] = `
DomainObjectMetadata {
  "decorations": Object {
    "unique": null,
    "updatable": null,
  },
  "extends": "DomainValueObject",
  "name": "Address",
  "properties": Object {
    "city": DomainObjectProperty {
      "nullable": false,
      "required": true,
      "type": "STRING",
    },
    "postal": DomainObjectProperty {
      "nullable": false,
      "required": true,
      "type": "STRING",
    },
    "state": DomainObjectProperty {
      "nullable": false,
      "required": true,
      "type": "STRING",
    },
    "street": DomainObjectProperty {
      "nullable": false,
      "required": true,
      "type": "STRING",
    },
    "suite": DomainObjectProperty {
      "nullable": true,
      "required": true,
      "type": "STRING",
    },
  },
}
`;
